### PR TITLE
Db export, again

### PR DESCRIPTION
**Pull request by danielballan**
This is a rebase of #76 with some extra commits to add a test and make it pass.

### CODE REVIEW

**Comment by danielballan** [comment]
This may be ready. It could use a careful read from more than one reviewer. As you can see, the outline for copying external files is here, but not finished. (The docs include warning about that.) I'd like to roll this out as is and handle the external files piece in a second pass.

File: databroker/broker.py [hunk 1]
```diff
@@ -708,6 +708,62 @@ def process(self, headers, func, fields=None, fill=False):
...
+            db.mds.insert_run_stop(**_clean_doct(header['stop']))
+            # insert fs
+            res_uids = self.get_resource_uids(header)
+            for uid in res_uids:
+                res = self.fs.resource_given_uid(uid)
+                db.fs.insert_resource(res['spec'],
+                                      res['resource_path'],
+                                      res['resource_kwargs'],
+                                      root=new_root)
+                                      # FIXME: revisit root when dealing
```
Comment (licode) [inline databroker/broker.py]: The new_root is the root place where the external data file is going to transferred? So basically we can transfer data between different servers.

**Comment by danielballan** [comment]
@licode Yes, that is the plan. As written so far, this only copies database documents between servers. A future iteration will make use of the new_root argument to copy files as well.

**Comment by licode** [comment]
Got it. thanks. 
Can we detect the root path automatically? So we do not need to manually change root, or keep this as an optional argument.

**Comment by danielballan** [comment]
How would that work?

Get Outlook for iOS<https://aka.ms/o0ukef>




On Thu, Sep 15, 2016 at 3:52 PM -0400, "Li Li" <notifications@github.com<mailto:notifications@github.com>> wrote:


Got it. thanks.
Can we detect the root path automatically? So we do not need to manually change root, or keep this as an optional argument.

**Comment by licode** [comment]
I think I oversimplify the issue. Please ignore that.

**Comment by licode** [comment]
👍  This looks good to me

File: databroker/broker.py [hunk 2]
```diff
@@ -708,6 +708,62 @@ def process(self, headers, func, fields=None, fill=False):
...
+                    db.mds.insert_event(**event)
+            db.mds.insert_run_stop(**_clean_doct(header['stop']))
+            # insert fs
+            res_uids = self.get_resource_uids(header)
+            for uid in res_uids:
+                res = self.fs.resource_given_uid(uid)
+                db.fs.insert_resource(res['spec'],
+                                      res['resource_path'],
+                                      res['resource_kwargs'],
+                                      root=new_root)
```
Comment (tacaswell) [inline databroker/broker.py]: This should pass through `res['root']` until this is actually implemented.

File: databroker/broker.py [hunk 3]
```diff
@@ -708,6 +708,62 @@ def process(self, headers, func, fields=None, fill=False):
...
+        """
+        try:
+            headers.items()
+        except AttributeError:
+            pass
+        else:
+            headers = [headers]
+        for header in headers:
+            # insert mds
+            db.mds.insert_run_start(**_clean_doct(header['start']))
```
Comment (tacaswell) [inline databroker/broker.py]: Why do you need this?

Comment (tacaswell) [inline databroker/broker.py]: and by 'this' I mean `_clean_doct`

File: databroker/broker.py [hunk 4]
```diff
@@ -799,3 +855,10 @@ def _munge_time(t, timezone):
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+
+def _clean_doct(d):
```
Comment (tacaswell) [inline databroker/broker.py]: There is a function in doct for doing this splitting.

File: databroker/broker.py (hunk 3 above)
Comment (danielballan) [inline databroker/broker.py]: Because a doct has a _name attribute hanging off of it that gets unpacked in insert_run_start.

Comment (danielballan) [inline databroker/broker.py]: I am pretty sure we should get rid of '_name' -- an artifact of mongoengine documents -- but that's a question for another day.

**Comment by danielballan** [comment]
This is close, but it requires a `FileStore` that implements resource_given_eid. @tacaswell only implemented this method on `FileStoreMoving`, which means that users would need to set up `FileStoreMoving` on the production databroker before exporting. It seem this is actually encouraging the opposite of careful behavior, but I think @tacaswell has some opinions about these methods, so I'll leave the rest to him.

**Comment by danielballan** [comment]
If https://github.com/NSLS-II/filestore/pull/121 is merged, this PR should just work. But I don't want to run roughshod over any concerns @tacaswell may have here, so please do not merge unless you are @tacaswell. :- )

File: databroker/core.py [hunk 5]
```diff
@@ -190,7 +191,7 @@ def get_events(mds, fs, headers, fields=None, stream_name=ALL, fill=False,
                 # Look in the descriptor, then start, then stop.
                 config_data_fields = set(filter(comp_re.match, config_data)) - selected_fields
                 for field in config_data_fields:
-                    selected_fields.append(field)
+                    selected_fields.add(field)
```
Comment (danielballan) [inline databroker/core.py]: This part is an unrelated bug fix.

**Comment by danielballan** [comment]
Note to whoever has the patience to finish this someday: `FileStore.datum_gen_given_resource` calls `core.get_datumkw_by_resuid_gen` which returns the kwargs from the datums but successfully obscures the datum_ids. FileStore is designed to make it hard to get this stuff, and that aspect of the design is proving to be a wild success.

File: databroker/broker.py (hunk 2 above)
Comment (danielballan) [inline databroker/broker.py]: Note to anyone following along: this is now actually implemented.

File: databroker/broker.py [hunk 6]
```diff
@@ -722,6 +724,73 @@ def process(self, headers, func, fields=None, fill=False):
...
+            res_uids = self.get_resource_uids(header)
+            for uid in res_uids:
+                fps = self.fs.copy_files(uid, new_root=new_root, **copy_kwargs)
+                file_pairs.extend(fps)
+                res = self.fs.resource_given_uid(uid)
+                db.fs.insert_resource(res['spec'],
+                                      res['resource_path'],
+                                      res['resource_kwargs'],
+                                      root=new_root)
+                datums = self.fs.datum_gen_given_resource(uid)
```
Comment (danielballan) [inline databroker/broker.py]: https://github.com/NSLS-II/filestore/pull/122 will make this method,`datum_gen_given_resource`, return the full datum document (including the ID which we need) instead of just the datum kwargs.

**Comment by danielballan** [comment]
Test passes locally and should pass here once https://github.com/NSLS-II/filestore/pull/122 and https://github.com/NSLS-II/bluesky/pull/612 are in. Phew.

**Comment by tacaswell** [comment]
Not wanting to feel left out I added a commit to this PR too!

**Comment by danielballan** [comment]
Test failure is unrelated to this PR but probably real. I'll open a separate issue and expand on that.

Seeing @tacaswell's marked approval above, and knowing that three of us have code in here, I guess I'll be the one to pull the trigger....

File: databroker/broker.py [hunk 7]
```diff
@@ -722,6 +724,75 @@ def process(self, headers, func, fields=None, fill=False):
 
     get_fields = staticmethod(get_fields)  # for convenience
 
+    def export(self, headers, db, new_root=None, copy_kwargs=None):
+        """
+        Export a list of headers.
+
+        .. warning::
```
Comment (CJ-Wright) [inline databroker/broker.py]: Does this warning still apply?

Comment (tacaswell) [inline databroker/broker.py]: It does not
